### PR TITLE
Feature: Engine support for Bays as Outfits

### PR DESCRIPTION
**Pull request by samrocketman**
Purpose
-------

These are engine only changes which add the possibility to create a `Drone Bay` or `Fighter Bay` outfit as a purchasable and installable outfit.

This enables `Fighter Bay` and `Drone Bay` outfits to be purchased from an outfitter and installed on gun or turret mounts.  This would convert Gun hardpoints and Turret hardpoints into fighter bays.

Download links
----------------

I maintain a custom fork of Endless Sky which includes all of my pull requests combined.  You can get the latest download of this fork here https://github.com/samrocketman/endless-sky/actions/runs/2266893824

Videos
--------

- [Bactrian super carrier](https://discord.com/channels/251118043411775489/947227869036286033/971055901735780463)
- [Fury drone miner](https://discord.com/channels/251118043411775489/947227869036286033/970583552234229790) (uses features from fleet mining pull request)
- [Mule mining carrier with 2 drones and 4 fighters](https://www.youtube.com/watch?v=XszYgI9CGRI)

Screen shots
--------------

Installed outfit can be moved around to different gun ports.

<details><summary>Click to expand and see screenshot</summary>

---

image

</details>

---

Navy carrier with drone bays and fighter bays installed in every gun port and turret mount.

<details><summary>Click to expand and see screenshot</summary>

---

image

</details>

---

Features added
--------------

This new feature adds the following:

- New `weapon` attributes `"Drone Bay"` and `"Fighter Bay"`
- `"Drone Bay"` outfits will dynamically generate Drone Bay hardpoints over the installed `Gun` weapon hardpoint.
- `"Fighter Bay"` outfits will dynamically generate Fighter Bay hardpoints over the installed `Turret` weapon hardpoint.
- Reorganizing where the bay is installed on a Gun or Turret will dynamically change where the docked fighter or drone is rendered externally on the ship.
- Drones mounted in guns always dock facing forward.
- Turret-based fighters now visually have new angles:
  - `0 degrees` angle for Fighter Bays mounted in the center turrets.
  - `45 degrees` angle for Fighter Bays mounted in starboard/right sided turrets.
  - `-45 degrees` angle for Fighter Bays mounted in port/left sided turrets.
- Outfit panel renders a proper description showing a Drone Bay or Fighter Bay is added when purchasing the outfit.

Related pull requests
---------------------

Related issues:

- https://github.com/endless-sky/endless-sky/issues/6396 requests external bays for fighters/drones.  This PR adds both `Drone Bay` and `Fighter Bay` outfit.  Both drones and fighters get mounted externally by this feature.
- https://github.com/endless-sky/endless-sky/issues/5549 mentions an unhappy scenario where drones were bought but fleet was unable to carry them.  This PR adds Drone Bay outfit.
- https://github.com/endless-sky/endless-sky/issues/4610 mentions having a generic "Bay" which could take a fighter or drone.  This change adding Bay outfits would reduce the need for complicated engine logic to avoid fighters/bays taking the wrong bay which would leave a mixed-bay fighter/drone stranded.  The playre can choose the bay they wish to have installed by the outfit being purchased.
- https://github.com/endless-sky/endless-sky/issues/4034 this change adds a cheap drone which is easy to replace.

### CODE REVIEW

File: source/Armament.cpp
Comment (quyykk) [inline source/Armament.cpp:168]: This doesn't really change anything.

File: source/Ship.cpp
Comment (samrocketman) [inline source/Ship.cpp:689]:
```suggestion
	// Add bays based on installed Drone Bay or Fighter Bay outfits.
```

File: source/Ship.cpp
Comment (samrocketman) [inline source/Ship.cpp:862]:
```suggestion
```

Orientation was fixed.  I got weird orientation issues because there was a memory leak in my unpublished change with regard to boundless appending of fighter and drone bays.  I added `ScrubWeaponBays()` method to clear bays when player loads save game or reloads.  I forgot to remove this comment.

File: source/Ship.cpp
Comment (samrocketman) [inline source/Ship.cpp:881]: `bayIndex` is necessary because some ships use the same X/Y coordinates for multiple guns.  One example is the Bastion which has multiple instances of 2 guns mounted on the same coordinates.  I added `bayIndex` for this very reason as a fix before publishing code.

File: source/Ship.cpp
Comment (samrocketman) [inline source/Ship.cpp:896]: Adding two points creates a new point instance.

File: source/Ship.cpp
Comment (samrocketman) [inline source/Ship.cpp:1362]:
```suggestion
```
Delete; unused experiment.

File: source/Ship.cpp
Comment (samrocketman) [inline source/Ship.cpp:1363]:
```suggestion
```
Delete; unused experiment.

File: source/Ship.cpp
Comment (samrocketman) [inline source/Ship.cpp:1364]:
```suggestion
```
Delete; unused experiment.

File: source/Ship.cpp
Comment (samrocketman) [inline source/Ship.cpp:1365]:
```suggestion
```
Delete; unused experiment.

File: source/Ship.cpp
Comment (samrocketman) [inline source/Ship.cpp:3842]: This should probably be behind a conditional check if outfit is a weapon and is a bay.

**Review by samrocketman — COMMENTED** [review]
Apply fixes

File: source/Ship.cpp
Comment (quyykk) [reply source/Ship.cpp:895]: C++ doesn't have implicit reference semantics like Java and C# do. You can simply assign the point and it'll make a copy.

**Review by quyykk — CHANGES_REQUESTED** [review]
Cool feature! My only request is to make this more customizable for content creators instead of hardcoding everything.

This means:

- Ability for converting hardpoints to custom bay categories
- Ability to specify which hardpoint type the outfit is for
- For gun hardpoints respect the "angle" property.

I think that's it.

**Comment by tehhowch** [comment]
This is an intentional design point of Endless Sky: bays are not outfits and cannot be added or removed on a whim. Why are we considering changing that?

**Comment by Amazinite** [comment]
I've stated that I'm not against the idea of adding functionality that would allow bay outfits to exist, but I don't think that it's something that should be used in vanilla. That is, maintain it as an intentional design point of Endless Sky the game that bays aren't outfits, but allow Endless Sky the engine to make such outfits.

File: source/Ship.cpp
Comment (samrocketman) [reply source/Ship.cpp]: Conditional added

File: source/Armament.cpp
Comment (samrocketman) [reply source/Armament.cpp:168]: This was a mistake.  You're right it was a failed experiment that I aborted.  I'll revert this.

File: source/Ship.cpp
Comment (samrocketman) [reply source/Ship.cpp:895]:
```suggestion
	bay.point = hardpoint.GetPoint();
```
Like this?

**Comment by Zitchas** [comment]
> This is an intentional design point of Endless Sky: bays are not outfits and cannot be added or removed on a whim. Why are we considering changing that?

While it may have originally been a design decision to not have the ability to add bays as outfits, it is something that has been highly requested. Even if having outfit bays in vanilla is something that has been "Noed," there's no reason why we should explicitly deny this flexibility to plug-in creators; and plenty of reasons why we should.

**Comment by samrocketman** [comment]
@quyykk 

> Ability for converting hardpoints to custom bay categories

I'm still new to the hardpoint code.  How would you see this as a feature?  Can you describe in more detail?  As-is I don't know what changes to make, if any.

> Ability to specify which hardpoint type the outfit is for

Is this in-game by a player?  Is this for plugin creators via properties in txt files?  I'm not sure what changes to make based on this request.

@tehhowch 

> This is an intentional design point of Endless Sky: bays are not outfits and cannot be added or removed on a whim. Why are we considering changing that?

I've been discussing this in discord for the past week.  Should I ping you more on topics such as these?  I feel a disconnect between core maintainers and my changes lately.  I don't want to work on stuff unlikely to be merged.  I split out the data files into a separate PR #6793 since it was unlikely to be merged after discussion with Derpy in discord.

With regard to this feature: it does not grant the player ability to arbitrarily remove bays.  If a ship comes with bays (drone or fighter) as part of its design those are considered static and not changeable.  However, if an outfit is installed the outfit itself does bay +1 on the weapon hardpoint where the bay is installed.  Those bays can be removed if they're added by an outfit.

**Comment by samrocketman** [comment]
@quyykk all code changes made upon your request except for the ones I didn't understand or have enough detail on.

@tehhowch @Amazinite adding on to my earlier comments and even pulling comments out of discord...

This is being added as an enhancement for gameplay around using fighters and drones.  The bays themselves (even if added as a plugin) do not come free to the player.  They're limited by the amount of weapon hardpoints available (and type of weapon). Here's a breakdown of compelling points:

- The carrier's balance or really any ships balance is not drastically affected by replacing all weapons with bays.  It just has more fighters/drones but significantly reduced combat power compared to a ship simply equipping weapons.
- (not in this change but planned) Fighters should use long-range weaponry if they have them first.  Fall back to short-range weaponry defensively.
- Long-range energy/ammo based weapons that could be installed on a fighter are end-game outfits except for homing missiles.
- Carrier-types still have a place because they can be more deadly than other ship types since they have built-in bays.  So can carry that much more.  These are basically fighters and drones for free from a player perspective.
- Up front cost for new players to use fighters or drones today is 7 million credits.  Bays reduce the up front cost to < 500k credits + flagship cost.
- The bays add a lot of mass paired with the mass of the fighters.  In practice, while play testing this... this makes large carriers with lots of bays sitting ducks because they're slow to turn and accelerate.  Play testing this feels pretty clean with the heavy tradeoffs.

**Review by Zitchas — DISMISSED** [review]
I can't speak to the quality of the code, but I have tested it (via #6793 ) and it works as advertised. I bought and sold bays, filled them up with ships, which I could launch and recover.

Works great! Seeing drones attached on the front of my guns looks a little silly, but having fighters latch on to hardpoints where my turrets  used to be looks awesome. Honestly, I think I'd rather have drones at least have the option to be internal, defined in the outfit description.

Anyway, thanks for working on this; I'm looking forward to seeing this in-game so plug-in makers can add a lot of interesting new capabilities to our ships!

**Comment by Hurleveur** [comment]
As of now, I never use fighters as they are totally pointless; they die in seconds and don't do enough damage to warrant the pain to have to remember to carry them back before jumping (and loss of time). 
They could be okay as swarm but the kind of money needed and what you get back for it just doesn't feel right (and any other swarm is better because it would have less losses).

These changes (with the mining PR) look honestly very good and could make a very limited use ship become *finally* useful, as in you could have a carrier in your fleet specialized into making fuel for your other ships, this also makes using small ships way more viable as you can refuel them easily (instead of waiting for ages for a big fuel tank ship to refuel them)

That said balance seems to be an issue here (as always?), and I feel that instead of weighting in this new possible weapon that is a fighter (an expensive missile of such) and new possible customisation (which for me is the biggest positive point of this game, with the story) some just want to avoid that much needed question of balance at all and remove this option from this game, instead of finding what weapon; outfit space, maybe even energy generation would be needed to make this balanced.
If that is indeed the case and there are no real arguments behind not allowing new content in vanilla, I strongly disagree with that.

**Comment by petervdmeer** [comment]
I'm also not against the idea that we would allow "bay providing outfits" to exist (at least in the engine to allow plugins to use those), but I have a fundamental concern with the limited scope of this PR.

If we want to allow Gunports to be convertable to Bays, then we should have something like a generic notion of a "mount point" and we should be able to specify per mount-point what would be the default "operating mode" and what "conversions we allow on each mount point".

An overall vision (think [RFC](https://github.com/EndlessSkyCommunity/endless-sky/blob/community/docs/rfcs/0000-template-rfc.md) ) on such "mount points" should also tell me how I convert Engine ports into backwards firing gun-ports (or even Bays) and it should tell me how a content-creator that is creating ships can control what conversions can happen on which "mount point".

A ship-content-creator should be able to prevent guns that are fully built-in (with just a small part of the barrel sticking out of the ship) from being converted to Bay/docking points.
The ship-content-creator should also be able to specify if a "mount point" conversion to a bay results in an "INSIDE", "OVER" or "UNDER" bay. (And for turrets the "mount point" should also give me the default bay angle.)
The overall vision should also tell me how those conversions affect engine-space, weapon space (and maybe also outfit space) and issue #4700 is also related to this PR.

If we allow "bay providing outfits", then I also want to have recursive bay support properly implemented (with a good set of automated tests to safeguard it). I should be able to install a "drone carry" outfit on a Boxwing; So a Drone docks into the Boxwing with my outfit, the Boxwing docks into a Aerie.)

I also agree to all of @quyykk remarks about customization, not hardcoding, and supporting all bay types. And a feature like this should also have an extensive set of automated tests that show that the feature works as intented, and that safeguard such a feature (with the tests acting as regression tests) from being broken in the future.

Not wanting to discourage you, but my opinion about the "mergability of this PR" is that it is not mergeable at least until:
- The overall vision is in place.
- The "affected system-functions" all have good regression tests.
- The new functionality has good regression tests.
- More basic functionality (like restricting missile launchers to some gunports and not others, and setting limits in outfit space on gunports itself) is in place.

I'm very happy with all the enthousiasm with which you are implementing new features. But IMHO there is quite a bit of rework on the core game-engine that needs to be done before we can get to very cool but very impactfull features like this.

**Comment by samrocketman** [comment]
@petervdmeer good to know an RFC process exists as an idea in ES.  I was not aware and probably would have followed it to begin with.  (I understand my current work has been gun ho).  With regard to fighter bays as outfits.  I'm open to making all of the changes (more modular customization, respecting angles, integration tests, etc.).  I have a question with regard to RFC feedback you shared.  Would you like me to put forth the vision/idea inside of the pull request itself as a comment?  Would you rather I file separate pull requests for RFCs and sponsor the RFC work as a dedicated pull request to be merged (RFC document)?
With regard to testing is there a metric I should be targeting?  Enough cases to be satisfied it won't break?  `100%` coverage, `>100%` coverage?

**Comment by petervdmeer** [comment]
@samrocketman I'm not deciding the course for the game alone; we have a group of about 7 maintainers that all could have an opinion on this and I notice that none of the other maintainers did respond to my latest comment yet.

Part of my comment was really content-creation related; like the question if we should allow each and every hardpoint to be convertible to a bay (or if we should put restrictions on that). If all other content-creators feel that it is a good idea to allow each and every hardpoint to be convertible to a bay today, then I'm willing to step back from my statement and then you can just proceed with the implementation here.

Regarding the tests;
- I just merged #6770, which is a pre-requisite for integration-testing non-vanilla plugin-only functionality (the alternative test universes). This doesn't help you today, but it does give you multiple options;
   - Either you wait for me to finish the universes functionality (and then write an integration test for bay-convertors using the universes functionality).
   - Or you try to implement the universes test-functionality yourself (I'm willing to upload a branch with half of the functionality in, and I don't have a lot of time the coming 2 weeks for ES, so you do have some time to try by yourself then.)
   - Or if there is consensus that this PR can be merged without tests in place (maybe because it is plugin only functionality so we care less), then you could also try to convince us that this is mergeable without additional tests.
- I also have #6502 open, which tries to remove the GameData dependency from the Ship class. The GameData dependency is one of the factors that seems to prevent ships from being unit-testable. There still is quite a bit of work to get all the opens handled to make Ships and AIs unit-testable, but also there you have the options of waiting, contributing or convincing that testing can be ignored.

Sorry if my feedback is frustrating; I know how disappointing it is if you implement a very cool feature and then some maintainer is comming up with arguments like "long term vision", maintainability and testability. But I feel those arguments are quite important to avoid some nasty "technical debt" in the future.

Regarding RFCs: If you look at https://github.com/EndlessSkyCommunity/endless-sky/pulls then you see 3 open PRs marked as RFC. (You will also see that they are quite old already, so asking you to write an RFC might also feel a bit unfair. I created #6801 which I hope will also improve that process.)

**Comment by samrocketman** [comment]
Okay so I see three action items for me:

- Update this PR with less hard coding and better options for content creators, respect both turret and gun angles.  Support more hardpoints than just gun/turret inside the ship.  This is something I plan to do before merge.
- Regarding testing: Review @petervdmeer  work on test reorganization and get some context around intent by reading through comments and references (a lot has been said so I'll need to review a couple times likely)
- Separate from this PR: I can put together an RFC in the mean time around test infrastructure taking in suggestions already written by @petervdmeer and @quyykk on this PR and in the-hub discord.

**Comment by samrocketman** [comment]
I will add that the feature of having installable bays has been pretty enjoyable to play with.  I am replaying the the game start to finish with my developed PRs and so far it is pretty nice.

**Comment by petervdmeer** [comment]
> * Separate from this PR: I can put together an RFC in the mean time around test infrastructure taking in suggestions already written by @petervdmeer and @quyykk on this PR and in the-hub discord.

No, the test-infrastructure can already be found at https://github.com/endless-sky/endless-sky/tree/master/tests . No need to create an RFC for the test-infrastructure.

**Comment by samrocketman** [comment]
> No need to create an RFC for the test-infrastructure.

Just following up here since I said so in discord.  I will not be writing an RFC (in case someone is only following PR).

**Comment by petervdmeer** [comment]
Converted to Draft; this PR has open review-comments from quyykk and from me that are not handled yet.
(Please feel free to convert back to non-Draft once the review-comments are handled.)

**Comment by tehhowch** [comment]
> I've stated that I'm not against the idea of adding functionality that would allow bay outfits to exist, but I don't think that it's something that should be used in vanilla.

If you're going to say no, you need to _actually say no_. Otherwise you're saying "yes" and also not giving the feature the necessary due diligence.

> Even if having outfit bays in vanilla is something that has been "Noed," there's no reason why we should explicitly deny this flexibility

The reason is simple: unnecessary complexity. Adding things because we can is not free; contributors must understand the feature, both intent and implementation, to make source code changes. Plugin authors will need to inspect their content to determine if this new functionality will break their existing ship designs and intended atmosphere. New content will need to be balanced against the existence of this functionality.

Unused features are by nature untested features, meaning we will silently break them. As Peter identifies, we can alleviate this somewhat with tests. But adding tests is also not free, testing each PR takes longer, tests must be maintained, etc.

> I've been discussing this in discord for the past week.

I haven't been on Discord since mid February; pinging me will not create the free time needed.

--

I have no interest in adding a complex mechanic that will not be used in the game that we create, that we will need to maintain in perpetuity, for no benefit to ourselves. If it gets decided that this is a mechanic to be supported by the engine, then it better be vanilla-ready.

**Comment by samrocketman** [comment]
@tehhowch to clarify, could you expand on what you mean by vanilla ready?  The change is stable and works.  There is some feedback per code review but once implemented it would be fully play tested and considered stable for merging.  I put a lot of work into making sure all of my PRs are stable, production-ready, features.

**Comment by petervdmeer** [comment]
@samrocketman I'm sorry to disappoint you, but this feature is far from vanilla ready.

I listed a number of things [in my earlier comment](https://github.com/endless-sky/endless-sky/pull/6792#issuecomment-1119036605) and that is just a start of what I feel is needed to get a little bit closer to making this PR vanilla ready.
Your last replies suggest to me that you didn't understand what tehhowch or me said. You will however need to do a lot more than just handling our review-comments for a PR like this; you will need to have a deep understanding of the concerns behind the reviews. (And I expect that at least 5 other PRs need to be implemented and merged as basis before we could even consider merging this PR.)

If you're up to the challenge, then I'm very much okay with you trying to make this PR (and the 5 other PRs) vanilla ready. Feel free to ask for clarification if any of my review-comments are unclear, but also keep in mind that the time the maintainers have is limited. (And all the time we spend on your PRs is not spent on other PRs.)

**Comment by samrocketman** [comment]
> I'm sorry to disappoint you, but this feature is far from vanilla ready.

I think you misunderstood me.  I understand there's feedback in this PR.  I address feedback when it comes in and I am fine with you converting it back to a draft.  I was mostly meaning the PRs are stable code-wise but I agree with the feedback given and that the changes will need to be made.  My main concern with vanilla-ready comments was it felt like my code stability and idea in general was called into question.

I am hopeful this would be merged at all given the conditions 1) all feedback is addressed and there's no outstanding feedback 2) all prerequisite pull requests are merged.  The vanilla-ready comments make me think otherwise.

I believe this is a feature which would be used by many even as a plugin.  I play the game with this today.  So I do not believe this would be an obscure code path which would see little use.

> And I expect that at least 5 other PRs need to be implemented and merged as basis before we could even consider merging this PR

Are these PRs that you're referring to ones that I have already opened and am shepherding them currently through feedback?  Or are these yet-to-be implemented PRs?  In which case I need a little more to go on if I'm to participate in implementing them.

> If you're up to the challenge, then I'm very much okay with you trying to make this PR (and the 5 other PRs) vanilla ready.

I'm hopeful that I am.  I try to be forthcoming with as much information as I can in the PR itself (both the description and the commits).  I also try to address the feedback either with comments or with code changes.

> Feel free to ask for clarification if any of my review-comments are unclear

Thanks I will; and in general I'm pretty good about asking others when I do not understand theirs.

> but also keep in mind that the time the maintainers have is limited. (And all the time we spend on your PRs is not spent on other PRs.)

I understand completely.  I am a maintainer in other large open source projects so I am sympathetic to these statements.  I try to approach all of the feedback in a positive light, with good intentions, and giving the benefit of the doubt.  Your feedback (tehhowch and others') has helped me become better at C++ and I value all of your time (being willing to share with me your experience).

In order to help alleviate maintainers, I have also taken your advice/request from discord to heart and have been both play testing and code reviewing other pull requests that are not my own in order to help facilitate their merging.

---

As an aside, I found bays as outfits to be quite useful in order to both find, fix, and test #6866.  The ability to add/remove bays to try out different scenarios made testing the fix quite easy than it otherwise would have without.

**Comment by samrocketman** [comment]
Let's set code implementation aside for a second and talk purely data files.  I want to implement this in a way that is flexible for content creators per the feedback.  Help me see how this would look or behave.  I'm thinking something like the following.

```groovy
outfit "Fighter Bay"
	category "Turrets"
	cost 320000
	thumbnail "outfit/unknown"
	"mass" 40
	"outfit space" -40
	"weapon capacity" -40
	"turret mounts" -1
	"required crew" 1
	"unplunderable" 1
	bay "Fighter"		//<-- Enables more generic flexibility beyond weapons.
	bay_angle = 45		//<-- Not adding angle means 0. by default or turret or gun angle if weapon bay is chosen.
	bay_position = (0,0)	//<-- I'm not sure how point should look here for where bay should be stored.
	weapon
		Bay
		// Any other weapon-specific attributes I'm forgetting?
```

**Comment by petervdmeer** [comment]
@samrocketman

> My main concern with vanilla-ready comments was it felt like my code stability and idea in general was called into question.

Your code stability was not called into question, but the code maintainability was (and still is).

The idea in general was also called into question, and that is actually a good thing; implementing some features can block implementing other features in the future, so we need to make well informed choices.

> I am hopeful this would be merged given ...

That's not how it works, not for this PR. If you continue working on this PR (possibly re-implementing it multiple times), then there still is a chance that in the end we will still find a reason not to merge this at all.
Tehhowch is one of the most experienced members in the community, so if tehhowch says that something is a bad idea, then consider that to be an early warning that all work on this PR might be in vain in the end.

> I believe this is a feature which would be used by many even as a plugin.

Drop the idea of plugins; this feature will need to be so well defined that an experienced content-creator would be able to use it in one of the ships in vanilla (without affecting any other ships). Anything less than that would be non-vanilla ready.

Which brings me to one of the 5 other PRs I was talking about; the PR that implements https://github.com/endless-sky/endless-sky/issues/5443 will likely be needed to avoid [the silly looking attached drones from Zitchas comment](https://github.com/endless-sky/endless-sky/pull/6792#pullrequestreview-961308003).

Btw, I found https://github.com/endless-sky/endless-sky/issues/3776 while looking for the other issue, 3776 seems related to this PR (or to one of the 5 other PRs I was talking about).

**Comment by samrocketman** [comment]
> if tehhowch expresses doubts about some idea, then consider that to be an early warning that all work on this PR might be in vain in the end.

That's fair and so be it if it comes to be.  I will continue to maintain the mod in my fork even if it never makes it into vanilla for the community.

I will play the game the way I want and update the code on the basis that it may get merged.

If this PR gets closed that will be the final indicator that this is full rejected and I won't raise a fuss.  I'll just maintain it for myself.

**Comment by samrocketman** [comment]
https://discordapp.com/channels/251118043411775489/308902312741568512/980224843859496970

Another idea: bay filter for the kind of ship that is allowed to be stored.

Kitteh:

> Could be it be feasible to design a heavy warship that during heavy combat can deply 'chunks' of it iself as light warships or interceptors to assist with the fight? How hard could it be to make it so only its designed 'fighters' dock with it?

You can filter what could dock.  Bays as outfits allows a leviathan to have any fighter.  But it would be cool to have fighter-specific bays.  Such as the mining drone bay (only holds a mining drone).

**Comment by samrocketman** [comment]
I went ahead and updated my example to include a ship filter in comment https://github.com/endless-sky/endless-sky/pull/6792#issuecomment-1138166615

**Comment by petervdmeer** [comment]
> Another idea: bay filter for the kind of ship that is allowed to be stored.

You're trying to cram too many different features into a single PR. This type of filter has been requested multiple times already, and should have it's own PR (for regular bays first) before we should even discuss it here.

**Comment by samrocketman** [comment]
> > Another idea: bay filter for the kind of ship that is allowed to be stored.
> 
> You're trying to cram too many different features into a single PR. This type of filter has been requested multiple times already, and should have it's own PR (for regular bays first) before we should even discuss it here.

I'll do what I want as long as it is fun.

I'm just playing the game and enjoying the bays.  I asked for feedback above on the **data file format** for this feature; it would be great if I could get feedback on it.  Currently, I've been asked to make changes yet no direct feedback on how those changes could be or what they should look like from an end-user's perspective.  I'm going to push this forward one way or another regardless of opinions expressed here.

The nice thing about a fork is you can do whatever you want.

---

> This type of filter has been requested multiple times already

Is there a GitHub issue?  Is it well defined and could be implemented by a contributor.  I could contribute something like this but not before my existing pull requests are addressed.

It sounds to me like the core devs and I have very different ideas on how we'd like to play the game.  At some point, this process has become painful for me so I'm going to focus more on what I enjoy and less on caring whether or not my features get accepted.

**Comment by petervdmeer** [comment]
I'm sorry to hear you don't feel welcome in the ES community. From my side I'm very happy with you being part of the community and I also noticed a lot of positive responses from senior community members like Amazinite, Quyykk, Zitchas and Hurleveur already just on this PR.
Tehhowch and me are also not actively opposing the feature itself, but it is more of a "if we are going to do this, then we better make sure it is very well done".

Many active developers that quickly want to try out something have some sort of (private) fork or private plug-in, so I fully understand that you are considering to do something similar.
If you want to contribute code to the vanilla main game, and feel comfortable to interact with the community, then you are also always welcome to create or update a PR here.

> Is there a GitHub issue?

Yes, #5443, #4684 and #4610 seem relevant.

I also noticed #692 and #3795 when looking through the issues; those seem relevant for this PR.

Note: I'm not going to spend time on providing feedback on the data file format, because I feel that [the overall vision](https://github.com/endless-sky/endless-sky/pull/6792#issuecomment-1119036605) first needs to be created. There is no point in discussing details when the bigger picture is still missing.

**Comment by samrocketman** [comment]
> I'm sorry to hear you don't feel welcome in the ES community.

It's not that I don't feel welcome; **I do**.  What I meant by painful (probably better described as impatience or angst) is the following.

The pace in which I would like to develop vs the pace of my PRs are merged has been a challenge to balance.  I would like to develop the AI further as my primary focus and the review cycle between changes has been fairly long.

I understand this is spare time for all of us but still pretty hard to balance sometimes for me.  Sometimes it's hard to control my excitement to develop further which is how the low powered fighter PR (and general fighter AI overhaul) has grown to be so large.

---

I appreciate you taking the time to explain the process and outline future-looking what you'd like to see.

**Comment by Hurleveur** [comment]
since you're asking about how it should be presented, I'll give my opinion on it
```cpp
outfit "Fighter Bay"
	category "Bay"			//this is so different it should not appear in turrets at all in my opinion, but under a new category, this could even be under system, but it isn't an actual turret
	cost 320000
	thumbnail "outfit/unknown"
	"mass" 40
	"outfit space" -40
	"weapon capacity" -40
	"turret mounts" -1
	"required crew" 1
	"unplunderable" 1
// Above were the details, now the real stuff
	// a bay definition like we can see for ships would be neat, idk it by heart and we should prob add things but still it should be quite similar (and could be in the code as well)
	// Like this as an easy option that takes all the default things (and is the same as the one for ships)
	bay "Fighter" -70 -45 left under

// But we could want to have a more developed version (that would be made optional to the ship as well) One thing is sure I don't want these attributes as they currently are, because they would stack on the ship and all. We need proper and special treatment for them. So, the more developped thing, that ships in general would be able to use as well (idk if they already can) would be smth of the like:
	bay		// This way we don't say "bay that" "bay this" later on
		category "Fighter"			//<-- Enables more generic flexibility beyond weapons.
		"angle" 45			//<-- Not adding angle means 0. by default or turret or gun angle if weapon bay is chosen.  Alternate angle could be "bow" or "stern" which sets 45 degrees like hard coding in PR.
		"position" 0 0		//<-- I'm not sure how point should look here for where bay should be stored.
		"side" over			//<-- defaults to weapon type (turret=over or gun=under) but can be overridden otherwise default is "inside"

		"ship" "Mining Drone"	//<-- Filter so that only a specific ships is allowed.  Could allow more than one.
		// Idk about this one above, but if we add bays this way we probably have to add an outfit size limit for the ships in it on top.

		hardpoint "weapon"			//<-- This indicates the bay should be a weapon type and so look to the hardpoint for bay position
We cannot really use the weapon thing here without losing on clarity (weapon in that definition is smth that happens when the ship dies, it could later on be expanded to be a subweapon or idk what, we can't use that in my opinion)
```

I think that the definition of the bay should match Bays in general in game. If we want to add stuff, like `"bay ship" "Mining Drone"`, lets do it so that all CC can use it on any ship, for any Bay.

**Comment by Hurleveur** [comment]
The thing is, this does not only add the possibility of bays as outfits, it also adds some new limits for those bays (why would it be them only) and as such, could be split into 2 parts, one that just adds more options for current bays, and the other one that is just allowing for a bay as outfit that follow the same syntax.

Basically what tehhowch was saying is this could be an improper way of doing it, I think it should be done via a class that sets up all bays, these and normal ones, and thus be quite small in complexity, hopefully?

**Comment by samrocketman** [comment]
Update: I still plan to do this rework.  I am focused on getting PR #6669 in the game and then following up with the battery powered AI overhaul refactor.  Once those two are in I'll refocus on this PR.  I haven't forgotten about this.

**Comment by Amazinite** [comment]
Closing this PR as the implementation is very off-base from where we want it to be. I understand that Sam is looking to refactor it once other PRs are taken care of, but that is also true of #6726. I don't want to have too many open draft PRs sitting around waiting on other PRs to be finished first, and comparing this PR to that one, I'd say this PR has a fair bit further to go to be where we want it to be, while #6726 mainly just needs various changes split out of it.

We can revisit this idea in the future once we have the time to, but for now, this PR can be closed.

**Comment by samrocketman** [comment]
Feedback for later refactor.  Provide an attribute option where externally docked fighters or drones can fire their weapons https://discord.com/channels/251118043411775489/431496424992014347/1071560352610918460

So they still act like ship guns.

**Pull request closed without merging**
